src/persist/src/indexed/mod.rs

```diff
@@ -292,63 +296,84 @@ impl AppliedState {
-            self.prev_meta = new_meta;
+            self.state.saved_seqno = new_meta.unsealeds_seqno_upper;
+            self.state.prev_meta = new_meta.clone();
         }
 
-        metrics
+        self.metrics
             .stream_count
-            .set(u64::cast_from(self.prev_meta.id_mapping.len()));
-        let unsealed_blob_count: usize = self
-            .prev_meta
-            .unsealeds
-            .iter()
-            .map(|x| x.batches.len())
-            .sum();
-        metrics
+            .set(u64::cast_from(new_meta.id_mapping.len()));
+        let unsealed_blob_count: usize = new_meta.unsealeds.iter().map(|x| x.batches.len()).sum();
+        self.metrics
             .unsealed_blob_count
             .set(u64::cast_from(unsealed_blob_count));
-        let unsealed_blob_bytes: u64 = self
-            .prev_meta
+        let unsealed_blob_bytes: u64 = new_meta
             .unsealeds
             .iter()
             .flat_map(|x| x.batches.iter().map(|x| x.size_bytes))
             .sum();
-        metrics.unsealed_blob_bytes.set(unsealed_blob_bytes);
-        let trace_blob_count: usize = self.prev_meta.traces.iter().map(|x| x.batches.len()).sum();
-        metrics
+        self.metrics.unsealed_blob_bytes.set(unsealed_blob_bytes);
+        let trace_blob_count: usize = new_meta.traces.iter().map(|x| x.batches.len()).sum();
+        self.metrics
             .trace_blob_count
             .set(u64::cast_from(trace_blob_count));
-        let trace_blob_bytes: u64 = self
-            .prev_meta
+        let trace_blob_bytes: u64 = new_meta
             .traces
             .iter()
             .flat_map(|x| x.batches.iter().map(|x| x.size_bytes))
             .sum();
-        metrics.trace_blob_bytes.set(trace_blob_bytes);
+        self.metrics.trace_blob_bytes.set(trace_blob_bytes);
 
         Ok(())
     }
-}
 
-impl<L: Log, B: Blob> Indexed<L, B> {
+    /// Applies an unbatched cmd to the machine state and snapshots the result
+    /// to durable storage.
+    ///
+    /// Precondition: pending has been emptied
+    fn apply_unbatched_cmd<
+        T,
+        WorkFn: FnOnce(&mut AppliedState, &mut BlobCache<B>, &mut Maintainer<B>) -> Result<T, Error>,
+    >(
+        &mut self,
+        work_fn: WorkFn,
+    ) -> Result<T, Error> {
+        debug_assert_eq!(self.validate(), Ok(()));
+        // NB: This validate_empty is intentionally a returned error instead of
+        // an assert because it's a precondition (and so a violation means a
+        // usage error by the caller of this).
+        self.pending.validate_empty()?;
+
+        let work_ret = match work_fn(&mut self.state, &mut self.blob, &mut self.maintainer) {
+            Ok(work_ret) => work_ret,
+            Err(err) => {
+                self.state.restore();
+                return Err(err);
+            }
+        };
+        self.try_set_meta()?;
+
+        debug_assert_eq!(self.validate(), Ok(()));
```

Comment:
I think checking that pending is empty is a good postcondition here as well?

Reply:
Done